### PR TITLE
Adding debug level to logging and conditionally logging messages

### PR DESCRIPTION
**Pull request by lukewaters**
between debug and verbose levels

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@lukewaters__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Luke Waters). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla2.msopentech.com.
    </p>

TTYL, MSOTBOT;

File: src/Microsoft.Framework.Logging.Serilog/SerilogLogger.cs [hunk 1]
```diff
@@ -89,6 +89,8 @@ private LogEventLevel ConvertLevel(LogLevel logLevel)
                 case LogLevel.Information:
                     return LogEventLevel.Information;
                 case LogLevel.Verbose:
+                    return LogEventLevel.Debug;
+                case LogLevel.Debug:
                     return LogEventLevel.Verbose;
                 default:
                     throw new NotSupportedException();
```
Comment (lukewaters) [inline src/Microsoft.Framework.Logging.Serilog/SerilogLogger.cs]: I looked at the log levels in 9 popular loggers, and serilog is the only logger that has debug as a lower level than verbose.  I am open to suggestions for as to what to do in this case

**Comment by lukewaters** [comment]
This addresses issues #103 and #104

**Comment by lukewaters** [comment]
Take a look @blowdart

**Comment by blowdart** [comment]
LogLevel.Gossip? :) Debug works for me, doc comment should note the worry that PII etc could be pumped out at this level.

File: src/Microsoft.Framework.Logging/LoggerExtensions.cs [hunk 2]
```diff
@@ -18,6 +18,131 @@ public static class LoggerExtensions
...
+            (state, error, secure) =>
+            {
+                var formatter = (ILoggerStructure[]) state;
+
+                if (secure)
+                {
+                    return LoggerStructureFormatter(formatter[0], error);
+                }
+                return LoggerStructureFormatter(formatter[1], error);
+
```
Comment (Eilon) [inline src/Microsoft.Framework.Logging/LoggerExtensions.cs]: Need to clean up lots of formatting (extra newline here; lots of other places are missing spaces).

File: src/Microsoft.Framework.Logging/LoggerExtensions.cs [hunk 3]
```diff
@@ -18,6 +18,131 @@ public static class LoggerExtensions
...
+        //------------------------------------------DEBUG------------------------------------------//
+
+        /// <summary>
+        /// Writes a verbose log message.
+        /// </summary>
+        /// <param name="logger">The <see cref="ILogger"/> to write to.</param>
+        /// <param name="data">The message to log at verbose level.</param>
+        /// <param name="secureData">The message to log at debug level.</param>
+        // FYI, this field is called data because naming it message triggers CA1303 and CA2204 for callers.
+        public static void WriteDebug([NotNull] this ILogger logger, string secureData, string data)
```
Comment (Eilon) [inline src/Microsoft.Framework.Logging/LoggerExtensions.cs]: Not really a fan of calling I "secureXyz" in various places. This is the only place that the security nomenclature seems to be used. Everywhere else it's a distinction between "debug" and not, and that seems orthogonal to the distinction between "secure" and not.

**Comment by davidfowl** [comment]
I dont understand why there an extra bool on the lambda. That looks completely misplaced. The only change that's supposed to happen was the addition of this Debug log level and a new `ILoggerFactory` property to set the minimum.

/cc @anurse @loudej

**Comment by Eilon** [comment]
Yeah this change seems to have a lot of unrelated things going on here that perhaps ought to be separated. (I know I missed some of the design discussions.)

**Comment by lodejard** [comment]
Yes, please submit a PR which adds the "Debug" level to the enum but does not change the formatting mechanics. The second step is to add an `ILoggerFactory.MinimumLevel` property which defaults to Verbose. The Logger instances at that point will check the level against the _factory.MinimumLevel in IsEnabled and WriteCore, and won't pass the call along to the providers' ILogger instances if the argument is lower than the property.

So, all sensitive information is logger at Debug, and to see it the app must set `loggerFactory.MinimumLevel = LogLevel.Debug` in code before it will flow.

**Comment by lukewaters** [comment]
I'm going to close this pull request as I clearly need to understand the problem better before I can fix the issue.
